File: scripts/check-duplication.mjs
import fs from 'node:fs/promises';

const html = await fs.readFile(process.argv[2] ?? 'docs/index.html', 'utf8');
const strip = (s) => s.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim().toLowerCase();

// ζεύγη: το ορατό κείμενο της κάρτας και το κείμενο του details της
const cards = html.split('<article').slice(1);
const problems = [];

cards.forEach((card, i) => {
  const det = card.match(/<details[\s\S]*?<\/details>/);
  if (!det) return;
  const hidden = strip(det[0]);
  const visible = strip(card.replace(det[0], ''));
  const hidWords = new Set(hidden.split(' ').filter(w => w.length > 4));
  if (!hidWords.size) return;
  let shared = 0;
  for (const w of hidWords) if (visible.includes(w)) shared++;
  const overlap = shared / hidWords.size;
  if (overlap > 0.5) problems.push(Κάρτα #: επικάλυψη %);
});

if (problems.length) {
  console.error('Το αναπτυσσόμενο μπλοκ επαναλαμβάνει το ορατό κείμενο:\n' + problems.join('\n'));
  process.exit(1);
}
console.log('Έλεγχος επανάληψης: καθαρό.');

File: scripts/check-duplication.py
import sys
import re

if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

filename = sys.argv[1] if len(sys.argv) > 1 else 'docs/index.html'
try:
    with open(filename, 'r', encoding='utf-8') as f:
        html = f.read()
except Exception as e:
    print(f"Error reading {filename}: {e}")
    sys.exit(1)

def strip(s):
    return re.sub(r'\s+', ' ', re.sub(r'<[^>]+>', ' ', s)).strip().lower()

cards = html.split('<article')[1:]
problems = []

for i, card in enumerate(cards):
    det_match = re.search(r'<details[\s\S]*?</details>', card)
    if not det_match:
        continue
    hidden = strip(det_match.group(0))
    visible = strip(card.replace(det_match.group(0), ''))
    hid_words = set(w for w in hidden.split(' ') if len(w) > 4)
    if not hid_words:
        continue
    shared = sum(1 for w in hid_words if w in visible)
    overlap = shared / len(hid_words)
    if overlap > 0.5:
        problems.append(f"Κάρτα #{i + 1}: επικάλυψη {overlap * 100:.0f}%")

if problems:
    print("Το αναπτυσσόμενο μπλοκ επαναλαμβάνει το ορατό κείμενο:\n" + "\n".join(problems))
    sys.exit(1)

print("Έλεγχος επανάληψης: καθαρό.")
